pq-crypto/s2n_kyber_evp.c

```diff
@@ -28,58 +28,123 @@
     POSIX_GUARD_PTR(kyber_pkey);
 
     DEFER_CLEANUP(EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new(kyber_pkey, NULL), EVP_PKEY_CTX_free_pointer);
     POSIX_GUARD_PTR(kyber_pkey_ctx);
 
-    size_t cipher_text_size = S2N_KYBER_512_R3_CIPHERTEXT_BYTES;
-    size_t shared_secret_size = S2N_KYBER_512_R3_SHARED_SECRET_BYTES;
-    POSIX_GUARD_OSSL(EVP_PKEY_encapsulate(kyber_pkey_ctx, ciphertext, &cipher_text_size, shared_secret,
+    POSIX_GUARD_OSSL(EVP_PKEY_encapsulate(kyber_pkey_ctx, ciphertext, &ciphertext_size, shared_secret,
                              &shared_secret_size),
             S2N_ERR_PQ_CRYPTO);
     return S2N_SUCCESS;
 }
 
-int s2n_kyber_512_evp_decapsulate(uint8_t *shared_secret, const uint8_t *ciphertext,
-        const uint8_t *secret_key)
+int s2n_kyber_evp_decapsulate(uint8_t *shared_secret, size_t shared_secret_size,
+        const uint8_t *ciphertext, const size_t ciphertext_size,
+        const uint8_t *secret_key, const size_t secret_key_size,
+        const int nid)
 {
-    size_t secret_key_size = S2N_KYBER_512_R3_SECRET_KEY_BYTES;
-    DEFER_CLEANUP(EVP_PKEY *kyber_pkey = EVP_PKEY_kem_new_raw_secret_key(NID_KYBER512_R3, secret_key, secret_key_size), EVP_PKEY_free_pointer);
+    DEFER_CLEANUP(EVP_PKEY *kyber_pkey = EVP_PKEY_kem_new_raw_secret_key(nid, secret_key, secret_key_size), EVP_PKEY_free_pointer);
     POSIX_GUARD_PTR(kyber_pkey);
 
     DEFER_CLEANUP(EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new(kyber_pkey, NULL), EVP_PKEY_CTX_free_pointer);
     POSIX_GUARD_PTR(kyber_pkey_ctx);
 
-    size_t shared_secret_size = S2N_KYBER_512_R3_SHARED_SECRET_BYTES;
     POSIX_GUARD_OSSL(EVP_PKEY_decapsulate(kyber_pkey_ctx, shared_secret, &shared_secret_size, (uint8_t *) ciphertext,
-                             S2N_KYBER_512_R3_CIPHERTEXT_BYTES),
+                             ciphertext_size),
             S2N_ERR_PQ_CRYPTO);
     return S2N_SUCCESS;
 }
+
+int s2n_kyber_512_evp_generate_keypair(uint8_t *public_key, uint8_t *secret_key)
+{
+    return s2n_kyber_evp_generate_keypair(public_key, S2N_KYBER_512_R3_PUBLIC_KEY_BYTES,
+            secret_key, S2N_KYBER_512_R3_SECRET_KEY_BYTES, NID_KYBER512_R3);
+}
+
+int s2n_kyber_512_evp_encapsulate(uint8_t *ciphertext, uint8_t *shared_secret,
+        const uint8_t *public_key)
+{
+    return s2n_kyber_evp_encapsulate(ciphertext, S2N_KYBER_512_R3_CIPHERTEXT_BYTES,
+            shared_secret, S2N_KYBER_512_R3_SHARED_SECRET_BYTES,
+            public_key, S2N_KYBER_512_R3_PUBLIC_KEY_BYTES, NID_KYBER512_R3);
+}
+
+int s2n_kyber_512_evp_decapsulate(uint8_t *shared_secret, const uint8_t *ciphertext,
+        const uint8_t *secret_key)
+{
+    return s2n_kyber_evp_decapsulate(shared_secret, S2N_KYBER_512_R3_SHARED_SECRET_BYTES,
+            ciphertext, S2N_KYBER_512_R3_CIPHERTEXT_BYTES,
+            secret_key, S2N_KYBER_512_R3_SECRET_KEY_BYTES, NID_KYBER512_R3);
```

Comment:
I wonder if you can simplify this logic to avoid all these wrapper methods. Maybe these methods should take s2n_kem as an argument and just read all the constants from the struct. Then most of your kems could just use s2n_kyber_evp_encapsulate / s2n_kyber_evp_decapsulate / s2n_kyber_evp_generate_keypair directly.

But maybe that's another follow-up issue.

Reply:
ah, i didn't realize that all of these constants are also stored in the `s2n_kem` struct, but it looks like they are:

```c
struct s2n_kem {
    const char *name;
    const kem_extension_size kem_extension_id;
    const kem_public_key_size public_key_length;
    const kem_private_key_size private_key_length;
    const kem_shared_secret_size shared_secret_key_length;
    const kem_ciphertext_key_size ciphertext_length;
    /* NIST Post Quantum KEM submissions require the following API for compatibility */
    int (*generate_keypair)(OUT uint8_t *public_key, OUT uint8_t *private_key);
    int (*encapsulate)(OUT uint8_t *ciphertext, OUT uint8_t *shared_secret, IN const uint8_t *public_key);
    int (*decapsulate)(OUT uint8_t *shared_secret, IN const uint8_t *ciphertext, IN const uint8_t *private_key);
};
```

I'll see if I can do as you suggest to cut down the LoC in this file.